File: python/src/x402_openai/core/networks.py
from __future__ import annotations

import re
from decimal import Decimal
from typing import Final, Literal, TypeAlias

BASE_MAINNET: Final = "eip155:8453"
BASE_SEPOLIA: Final = "eip155:84532"
SupportedNetwork: TypeAlias = Literal["eip155:8453", "eip155:84532"]

USDC_BY_NETWORK: Final[dict[SupportedNetwork, str]] = {
    BASE_MAINNET: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
    BASE_SEPOLIA: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
}

_USD_PATTERN = re.compile(r"^\d+(?:\.\d{1,6})?$")


def parse_usd_to_atomic(value: str) -> int:
    normalized = value.removeprefix("$")
    if not _USD_PATTERN.fullmatch(normalized):
        raise ValueError(f"Invalid USD amount: {value}")
    return int(Decimal(normalized) * 1_000_000)

File: typescript/packages/core/src/networks.ts
export const BASE_MAINNET = "eip155:8453" as const;
export const BASE_SEPOLIA = "eip155:84532" as const;

export const SUPPORTED_NETWORKS = [BASE_MAINNET, BASE_SEPOLIA] as const;
export type SupportedNetwork = (typeof SUPPORTED_NETWORKS)[number];

export const USDC_BY_NETWORK: Readonly<Record<SupportedNetwork, `0x${string}`>> = {
  [BASE_MAINNET]: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
  [BASE_SEPOLIA]: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
};

export function isSupportedNetwork(value: string): value is SupportedNetwork {
  return SUPPORTED_NETWORKS.some((network) => network === value);
}

/** Convert a decimal USD/USDC value to six-decimal atomic units without binary floats. */
export function parseUsdToAtomic(value: string): bigint {
  const normalized = value.startsWith("$") ? value.slice(1) : value;
  if (!/^\d+(?:\.\d{1,6})?$/.test(normalized)) {
    throw new Error(`Invalid USD amount: ${value}`);
  }

  const [whole = "0", fraction = ""] = normalized.split(".");
  return BigInt(whole) * 1_000_000n + BigInt(fraction.padEnd(6, "0"));
}
